costmap_2d/plugins/obstacle_layer.cpp

```diff
@@ -444,11 +448,58 @@ void ObstacleLayer::updateBounds(double robot_x, double robot_y, double robot_ya
   updateFootprint(robot_x, robot_y, robot_yaw, min_x, min_y, max_x, max_y);
 }
 
+void ObstacleLayer::computeFootprintCenter(const std::vector<geometry_msgs::Point>& footprint)
+{
+    double sum_x = 0.0, sum_y = 0.0;
+    for (const auto& point : footprint)
+    {
+        sum_x += point.x;
+        sum_y += point.y;
+    }
+
+    footprint_center_.x = sum_x / footprint.size();
+    footprint_center_.y = sum_y / footprint.size();
+}
+
+std::vector<geometry_msgs::Point> ObstacleLayer::getReducedFootprint(const std::vector<geometry_msgs::Point>& original_footprint, double reduction_size)
+{
+    std::vector<geometry_msgs::Point> reduced_footprint;
+
+    for (const auto& p : original_footprint)
+    {
+        double dx = p.x - footprint_center_.x;
+        double dy = p.y - footprint_center_.y;
+
+        double distance = sqrt(dx * dx + dy * dy);
+        geometry_msgs::Point new_point;
+
+        if (distance > 0)
+        {
+            double scale = (distance - reduction_size) / distance;
+            new_point.x = footprint_center_.x + dx * scale;
+            new_point.y = footprint_center_.y + dy * scale;
+        }
+        else
+        {
+            new_point = footprint_center_;
+        }
+
+        reduced_footprint.push_back(new_point);
+    }
+    return reduced_footprint;
```

Comment:
R: I think you should save the reduced footprint in the dynamic reconfigure call, so the computation of the new one is called only when you change the parameter (i.e, make this a member variable and in `updateFootprint` get it)